utils/sample_qc.py

```diff
@@ -0,0 +1,487 @@
+            max_y_female_call_rate=max_y_female_call_rate,
+            min_y_male_call_rate=min_y_male_call_rate
+        )
+    )
+
+    sex_ht =  sex_ht.drop('qc_platform')
+    return (sex_ht)
+
+
+def filter_duplicate_samples(
+        relatedness_ht: hl.Table,
+        samples_rankings_ht: hl.Table,
+        rank_ann: str = 'rank'
+):
+    """
+    Creates a HT with duplicated samples sets.
+    Each row is indexed by the sample that is kept and also contains the set of duplicate samples that should be filtered.
+
+    `samples_rankings_ht` is a HT containing a global rank for each of the sample (smaller is better).
+
+    :param Table relatedness_ht: Input relatedness HT
+    :param Table samples_rankings_ht: HT with global rank for each sample
+    :param str rank_ann: Annotation in `samples_ranking_ht` containing each sample global rank (smaller is better).
+    :return: HT with duplicate sample sets, including which to keep/filter
+    :rtype: Table
+    """
+    logger.info("Getting duplicate samples")
+    dups = get_duplicated_samples(relatedness_ht)
+    logger.info(f"Found {len(dups)} duplicate sets.")
+    dups_ht = hl.Table.parallelize([hl.struct(dup_set=i, dups=dups[i]) for i in range(0, len(dups))])
+    dups_ht = dups_ht.explode(dups_ht.dups, name='_dup')
+    dups_ht = dups_ht.key_by(**dups_ht._dup)
+    dups_ht = dups_ht.annotate(rank=samples_rankings_ht[dups_ht.key][rank_ann])
+    dups_cols = hl.bind(
+        lambda x: hl.struct(
+            kept=x[0],
+            filtered=x[1:]
+        ),
+        hl.sorted(hl.agg.collect(hl.tuple([dups_ht._dup, dups_ht.rank])), key=lambda x: x[1]).map(lambda x: x[0])
+    )
+    dups_ht = dups_ht.group_by(dups_ht.dup_set).aggregate(
+        **dups_cols
+    )
+
+    dups_ht = dups_ht.key_by(**{f'{x}_kept': expr for x, expr in dups_ht.kept.items()}).drop('kept')
+    return dups_ht
+
+
+def run_pca_with_relateds(
+        qc_mt: hl.MatrixTable,
+        related_samples_to_drop: Optional[hl.Table],
+        n_pcs: int = 10,
+        autosomes_only: bool = True
+) -> Tuple[List[float], hl.Table, hl.Table]:
+    """
+    First runs PCA excluding the given related samples,
+    then projects these samples in the PC space to return scores for all samples.
+
+    The `related_samples_to_drop` Table has to be keyed by the sample ID and all samples present in this
+    table will be excluded from the PCA.
+
+    The loadings Table returned also contains a `pca_af` annotation which is the allele frequency
+    used for PCA. This is useful to project other samples in the PC space.
+
+    :param MatrixTable qc_mt: Input QC MT
+    :param Table related_samples_to_drop: Optional table of related samples to drop
+    :param int n_pcs: Number of PCs to compute
+    :param bool autosomes_only: Whether to run the analysis on autosomes only
+    :return: eigenvalues, scores and loadings
+    :rtype: (list of float, Table, Table)
+    """
+
+    unrelated_mt = qc_mt.persist()
```

Comment:
Out of curiosity, what's the reasoning behind doing a `persist()` right off the bat?

Reply:
In this case, I believe the `persist()` is useful:
- We make some calculations on `unrelated_mt` that are computed from `qc_mt`
- then later we do `related_mt = qc_mt.filter_cols(hl.is_defined(related_samples_to_drop[qc_mt.col_key]))`

So the `persist()` indicates that `qc_mt` should stay in memory/disk so as not to be recomputed for the 2nd statement.